### PR TITLE
change server domain to 0.0.0.0

### PR DESCRIPTION
**Pull request by tiezhuli001**
change server domain to 0.0.0.0. allow remote ssh connections to access the http address via the ip:port format.

when the host is set to 0.0.0.0 in net.Listen(), it supports both IPv4 and IPv6. Refer to: https://github.com/golang/go/issues/22811

### CODE REVIEW

**Comment by ccoVeille** [comment]
I might be wrong but this change may lead that anyone on the same local network will be able to access your port, while it was not available with localhost.

As I said, I might be wrong, but it's worth trying. And if so, maybe it would be interesting to make this available via a command-line parameters (like --public or whatever)

**Comment by xhd2015** [comment]
The default behavior is to follow an npm package called `http-server`, which defaults to listen on public address. 
Since this tool is primarily used in local development, it's better to listen on local address by default. And can be made public by specifying `--public`. 

Goo point.

**Comment by tiezhuli001** [comment]
You're right. With regards to security, using the --public parameter certainly seems more prudent. The default is still set to the [localhost](https://localhost/):port configuration, but I'll make the appropriate changes. Thank you for your advice.

**Comment by ccoVeille** [comment]
> You're right. With regards to security, using the --public parameter certainly seems more prudent. Thank you for your advice.

I faced that issue when implementing something like that on a tool, that's why I raised the point.

Also, @xhd2015 @tiezhuli001 most people won't need the feature to listen on all network ever. So yes, it should be something to pass as argument

By the way, `--public` was a suggestion, a better name should be found I think.

**Comment by tiezhuli001** [comment]
Yes, thanks for the suggestions. Similar to Python's http.server and Node.js's http-server, I've added --bind and --port as options to set IP and port. The default host remains as [localhost](https://localhost/).

**Comment by ccoVeille** [comment]
Perfect, thanks

File: support/netutil/netutil.go [hunk 1]
```diff
@@ -72,3 +72,64 @@ func watchSignalWithinTimeout(timeout time.Duration, errSignal chan struct{}, ac
...
+	localIp := GetLocalHost()
+	validHosts := map[string]bool{
+		"localhost": true,
+		"0.0.0.0":   true,
+		localIp:     true,
+		"127.0.0.1": true,
+	}
+	if _, ok := validHosts[bindStr]; ok {
+		host = bindStr
+	}
```
Comment (ccoVeille) [inline support/netutil/netutil.go:112]: So no warning, error reporting here that you will fallback to local host?

File: support/netutil/netutil.go [hunk 2]
```diff
@@ -72,3 +72,64 @@ func watchSignalWithinTimeout(timeout time.Duration, errSignal chan struct{}, ac
...
+
+	if portStr == "" {
+		port = 7070
+	} else {
```
Comment (ccoVeille) [inline support/netutil/netutil.go:116]: What about removing the else by using return in the `if` above

File: support/netutil/netutil.go [hunk 3]
```diff
@@ -72,3 +72,64 @@ func watchSignalWithinTimeout(timeout time.Duration, errSignal chan struct{}, ac
...
+		if err != nil {
+			port = 7070
+		}
+		port = int(parsePort)
+	}
+	return
+}
+
+// construct URL based on host and port
+func GetAndDisplayUrl(host string, port int) string {
```
Comment (ccoVeille) [inline support/netutil/netutil.go:127]: Go initialisms or convention would enforce using GetAndDisplayURL (URL in uppercase as it's an acronym) here

File: support/netutil/netutil.go [hunk 4]
```diff
@@ -72,3 +72,64 @@ func watchSignalWithinTimeout(timeout time.Duration, errSignal chan struct{}, ac
 	}
 	action()
 }
+
+// get the local area network IP address
+func GetLocalHost() string {
```
Comment (ccoVeille) [inline support/netutil/netutil.go:77]: You are getting the local public IPs I would expect this method to return a list of string

I wan have eth0, tun0 and wlan0

You will return only the first one, so you cannot bind on the second or later ones

File: support/netutil/netutil.go [hunk 5]
```diff
@@ -72,3 +72,64 @@ func watchSignalWithinTimeout(timeout time.Duration, errSignal chan struct{}, ac
...
+			if ipNet.IP.To4() != nil {
+				return ipNet.IP.String()
+			}
+			// IPv6
+			if ipNet.IP.To16() != nil && ipNet.IP.To4() == nil {
+				return ipNet.IP.String()
+			}
+		}
+	}
+	fmt.Printf("no network interface found")
```
Comment (ccoVeille) [inline support/netutil/netutil.go:96]: This one may cause problem

If this method failed to find an IP, you will display an error message, but the user may have provided 0.0.0.0.

So you will use it, but an error will be displayed

It's unlikely to happen I think, but I would expect the local IPs to be tested after the valid one would be tested first

File: support/netutil/netutil.go [hunk 6]
```diff
@@ -72,3 +72,64 @@ func watchSignalWithinTimeout(timeout time.Duration, errSignal chan struct{}, ac
...
+
+// provide default values for host and port
+func GetHostAndIP(bindStr string, portStr string) (host string, port int) {
+	host = "localhost"
+	localIp := GetLocalHost()
+	validHosts := map[string]bool{
+		"localhost": true,
+		"0.0.0.0":   true,
+		localIp:     true,
+		"127.0.0.1": true,
```
Comment (ccoVeille) [inline support/netutil/netutil.go:108]: "::1" also no?

You are only considering IPv4 binding

File: support/netutil/netutil.go [hunk 7]
```diff
@@ -72,3 +72,64 @@ func watchSignalWithinTimeout(timeout time.Duration, errSignal chan struct{}, ac
...
+	}
+	return
+}
+
+// construct URL based on host and port
+func GetAndDisplayUrl(host string, port int) string {
+	url := fmt.Sprintf("http://%s:%d", host, port)
+	fmt.Println("Server listen at:")
+	fmt.Printf("-  Local: http://%s:%d \r\n", host, port)
+	if host == "0.0.0.0" {
```
Comment (ccoVeille) [inline support/netutil/netutil.go:131]: I don't think this is valid hereif someone provides a bind to his public IP you won't display it.

You will have to create a function to test if the bind is purely local I think and use icon other method, or copy paste the code here

**Review by ccoVeille — CHANGES_REQUESTED** [review]

File: support/netutil/netutil.go (hunk 1 above)
Comment (tiezhuli001) [reply support/netutil/netutil.go]: Yes, I agree that there's no need to return an error and terminate the process here. 
Regardless of the situation, we should open to the default host and port - [localhost](https://localhost/) should always be accessible, right?

File: support/netutil/netutil.go (hunk 3 above)
Comment (tiezhuli001) [reply support/netutil/netutil.go]: Yes, I need to modify my code. Thank you for your reminder.

File: support/netutil/netutil.go (hunk 4 above)
Comment (tiezhuli001) [reply support/netutil/netutil.go]: Yes, it currently supports passing in ipv4, ipv6 to get the corresponding list of IPs. Thank you.

File: support/netutil/netutil.go (hunk 6 above)
Comment (tiezhuli001) [reply support/netutil/netutil.go]: I didn't consider ipv6, I will try to fix it. Thank you.

**Comment by xhd2015** [comment]
Need to rebase master and squash to one commit

**Comment by tiezhuli001** [comment]
Yes, it has been merged into one commit, thank you.

File: support/netutil/netutil.go (hunk 7 above)
Comment (ccoVeille) [reply support/netutil/netutil.go:148]: Indeed, it only displays something if not listening on everything

**Comment by ccoVeille** [comment]
Thanks for applying the suggestions I made

**Pull request merged**